sklearn/model_selection/_validation.py

```diff
@@ -492,6 +493,43 @@ def _fit_and_score(estimator, X, y, scorer, train, test, verbose,
     return ret
 
 
+def _score_single_metric(estimator, X_test, y_test, scorer):
+    """Copmute the score of an estimator given a test set and a single metric.
+    This function will also check whether the estimator has a classes attribute
+    and pass that into the labels attribute of the scorer if available.
+    """
+    # check if estimator has a classes_ attribute and scorer has labels attr:
+    est_has_classes = hasattr(estimator, "classes_")
+    if hasattr(scorer, "_score_func"):
+        scorer_has_labels = "labels" in signature(
+                                        scorer._score_func).parameters
+    else:
+        scorer_has_labels = False
+
+    if est_has_classes & scorer_has_labels:
+        labels = getattr(estimator, "classes_", None)
+        scorer._set_score_func_parameters(labels=labels)
```

Comment:
@jnothman - this will overwrite in case the classes are set at the time of initialization. otherwise they'll anyways be inferred from y.

if the scorer already has a labels argument set, then we should probably check that the one passed in the estimator is the exact same as that of the scorer. if not, we should return an error. thoughts?

@amueller - with string input, check_scoring works as a lookup in a predefined dictionary. each object in the dictionary is a simple make_scorer(metric callable from API) without any additional parameters. we can modify it to a function call where we can pass the kwargs to pass on to the scorer, which I think is better way as well. But it looks like a significant change. if you guys are on board, I can go for it.